apps/internal/local/server.go

```diff
@@ -142,11 +164,23 @@ func (s *Server) handler(w http.ResponseWriter, r *http.Request) {
 
 	headerErr := q.Get("error")
 	if headerErr != "" {
-		desc := html.EscapeString(q.Get("error_description"))
 		// Note: It is a little weird we handle some errors by not going to the failPage. If they all should,
 		// change this to s.error() and make s.error() write the failPage instead of an error code.
-		_, _ = w.Write([]byte(fmt.Sprintf(failPage, headerErr, desc)))
-		s.putResult(Result{Err: fmt.Errorf(desc)})
+
+		errDesc := q.Get("error_description")
+
+		if bytes.Contains(s.errorPage, code) {
+			s.errorPage = bytes.Replace(s.errorPage, code, []byte(html.EscapeString(headerErr)), 1) // provides XSS protection
+		}
```

Comment:
1. Contains is just wasting time because Replace is a no-op when the slice doesn't contain the target
1. Replace returns a new slice you assign to `s.errorPage`. This makes the output of the each Replace the input to the next Replace. A Replaced slice has at least one fewer instance of the template argument. So, the net effect is that each argument replacement is permanent and once all arguments have been Replaced, all subsequent error pages are the same
1. you're replacing only the first instance of the argument. Why not replace all of them?

```suggestion
		errorPage := bytes.Replace(s.errorPage, code, []byte(html.EscapeString(headerErr)), -1) // provides XSS protection
```

Reply:
Hi @chlowell thanks for the feedback. I will address these a bit later (maybe tonight/tomorrow), but will get to it.

Reply:
Hi @chlowell I have made some more changes, please see my last 3 commits:

1: using Replace all
2: updated tests, using an `expected` value defined in the test definition
3: using type list syntax

Reply:
You haven't fixed the bug: https://play.golang.com/p/0I-sHtF-n83. Please also simplify the test. If each test case carries its expected content, the test body doesn't need a pile of conditional blocks to validate the actual content. It can simply check whether the actual content equals the expected content

Reply:
Hi @chlowell, can you please check again:
1: I have updated tests, removing the conditional logic. It's much simpler.
2: Updated the error page handling. I am no longer storing updated `errorspage` in the server struct.

Reply:
@chlowell, are free to take some time to review again please, and test ?